### ISSUE TITLE
include shortcode in computation code cell are not resolved for julia engine

### ISSUE THREAD

**Issue by cderv** [quarto-dev/quarto-cli#10008]
Following up on #9984 

Sorry, but I think I was confused by all the different things I tried.
Actually there seems to be a bug when using includes in the julia engine executable code block.

`test.qmd`

`````qmd
---
title: test
engine: julia
format:
  gfm: default
---

````{julia}
#| eval: false
{{< include foo.jl >}}
````
`````

`foo.jl`

`````julia
"""
```python
1 + 1
```
"""
`````

output:

`````md
# test


``` julia
"""
```python
1 + 1
```

““”

\`\`\`
`````

If it is not an executable code block, it works fine.

`````md
# test


```` julia
#| eval: false
"""
```python
1 + 1
```
"""
````
`````

So I suppose this is a bug of `QuartoNotebookRunner.jl`?
(The fact that the number of backticks varies from engine to engine was quite an unexpected behavior for me, though.)

_Originally posted by @eitsupi in https://github.com/quarto-dev/quarto-cli/issues/9984#issuecomment-2167027874_

**Comment by cderv** [comment]
I believe this is a cell source output backticks problem. Here is the intermediate outputs:
`````markdown
---
title: test
engine: julia
keep-md: true
format:
  gfm: default
---


::: {.cell execution_count=0}
``` {.julia .cell-code}
"""
```python
1 + 1
```
"""

```
:::

`````

We can see the the fenced code cell for source output does use only ```` backticks when initially it was 4. 

I guess not all engine handle nested backticks for code cells the same. Especially because it may be unexpected to have those. 
This is not valid Julia code right ? 😉 

I know **knitr** handles it for R by computing the number of backticks. No problem with `engine: knitr`

and it seems for Jupyter engine we do handle it also as I can't reproduce with this 

`````markdown
---
title: test
keep-md: true
format:
  gfm: default
---

````{python}
#| eval: false
{{< include foo.jl >}}
````

`````

So this is linked to new Julia engine that needs to do the same as with Jupyter engine. It may be Quarto and not `QuartoNotebookRunner.jl` directly.

**Comment by cderv** [comment]
Yes we do count backticks  at 
https://github.com/quarto-dev/quarto-cli/blob/50b4cf15a0be98a439773ffe923e1e284301879b/src/core/jupyter/jupyter.ts#L1422-L1427

**Comment by cderv** [comment]
It seems this issue does relate to shortcode because this has no issues

`````markdown
---
title: test
engine: julia
keep-md: true
format:
  gfm: default
---

````{julia}
#| eval: false
"""
```python
1 + 1
```
"""
````
`````

and produced correct intermediate md 

`````markdown
---
title: test
engine: julia
keep-md: true
format:
  gfm: default
---


::: {.cell execution_count=0}
```` {.julia .cell-code}
"""
```python
1 + 1
```
"""
````
:::


`````` 

So just maybe something about timing for resolving shortcodes ... 🤔

**Comment by eitsupi** [comment]
Thank you for your investigation!
The issue seems to be more complicated than I imagined...

**Comment by cderv** [comment]
OK so looking more into it, I do think this is not supported at the moment. 

@jkrumbiegel @cscheid as you were the one working on it in 
* https://github.com/quarto-dev/quarto-cli/pull/8645

It seems that Julia engine will use the input source file 
https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/execute/julia.ts#L120-L126

However, our internal processing does create a processed Markdown version of a .qmd file to resolve different things. And this includes shortcode. We resolves includes (more generally cell handler) in a pre-engine phase.

https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/command/render/render-contexts.ts#L274-L297

`context.target.markdown` is what will contains the resolved Markdown string that should be used for engine execution. 

This transformed markdown is used in several places like engine resolution
https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/execute/engine.ts#L167-L187

and used in engine execution. For example for Jupyter engine, we create a .ipynb from a .qmd 
https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/execute/jupyter/jupyter.ts#L291-L300
and we use this `target.markdown.value` and not the input .qmd file 
https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/execute/jupyter/jupyter.ts#L648-L655

So I believe julia engine should also use this markdown value for its engine execution. It should somehow be passed to QuartoNotebookRunner.jl 

- Either with an intermediate file 
	- like we do for the notebook creation where we create a `.quarto_ipynb` intermediate for jupyter engine
		https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/execute/jupyter/jupyter.ts#L201-L211
	- or with knitr engine where the markdown content from Quarto is passed to R and written to a file 
		https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/execute/rmd.ts#L123-L130
		https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/resources/rmd/execute.R#L9-L30
		
I don't know if QuartoNotebookRunner.jl requires only a file a input, or could work also with a markdown passed a string. 

but the same logic would need to be added. 

Also I do think this would solve our embed support, which rely also on this transformed markdown content as it is in the handler category;
- #8472 

Hope this helps know what to do exactly

**Comment by cderv** [comment]
> QuartoNotebookRunner takes a filepath as input, so we'd have to write out a temporary file I think.

I think that could be possible to write an intermediate file that would be removed after execution. 

Is this is strong requirement that QuartoNotebookRunner takes a file ? 

Asking because it could either be Quarto julia engine code writing it and removing it (like with Jupyter engine) , or otherwise this would be the QuartoNotebookRunner.jl side if we pass the information encoded as Json string (like with the knitr engine). It could even be base64 encode to be decode on Julia side for easier exchange of information. 

Not sure what is best. 

Do you want to work on this and send a PR for review ? Or should we add support for this (on the Quarto side in that case) ?

**Comment by jkrumbiegel** [comment]
Let me pull in @MichaelHatherly. We already get the full options JSON object over the wire, and I think this even contains the complete source as a string if I remember correctly. I assume some string in that object would change if shortcodes were executed correctly, and then we could pick it up from there?

**Comment by MichaelHatherly** [comment]
> We already get the full options JSON object sent over from quarto to QuartoNotebookRunner, and I think this even contains the complete source as a string if I remember correctly. I assume some string in that object would change if shortcodes were executed correctly, and then we could pick it up from there?

Yeah, we could just parse from that string in the JSON blob if it's available, otherwise fallback to the file.

**Comment by cderv** [comment]
> Let me pull in @MichaelHatherly. We already get the full options JSON object sent over from quarto to QuartoNotebookRunner, and I think this even contains the complete source as a string if I remember correctly. I assume some string in that object would change if shortcodes were executed correctly, and then we could pick it up from there?

Yes should have the `target.markdown` in there already on Quarto side. This is the markdown string where shortcodes like include and embed would have been expanded already. So if you pass everything to the Julia process already, you probably have access to it already. 

> Yeah, we could just parse from that string in the JSON blob if it's available, otherwise fallback to the file.

AFAIK `target.markdown` should always be set to the content of the file we read for all our processing.

**Comment by eitsupi** [comment]
@jkrumbiegel Thanks for the quick update, but I tried QuartoNotebookRunner 0.11.2 and it does not seem fixed.

**Comment by jkrumbiegel** [comment]
@eitsupi do you have an MWE? QuartoNotebookRunner just uses the preprocessed markdown from quarto now if it's available, so the question would be what side a possible bug would be on.

**Comment by eitsupi** [comment]
@jkrumbiegel Thanks for your quick reply. What does "MWE" mean?
The reproducible example in the first post of this issue does not work.
I am not sure if this is a Quarto CLI issue or a QuartoNotebookRunner issue, sorry. (I thought that a fix on the engine side only would fix this, but I could be mistaken.)

**Comment by mcanouil** [comment]
MWE: Minimal Working Example

**Comment by jkrumbiegel** [comment]
This is the output I get from the first post here, on QNB 0.11.2, with the latest prerelease `1.5.49`. Seems correct to me?

`````
# test


```` julia
"""
```python
1 + 1
```
"""
````
`````

**Comment by eitsupi** [comment]
Thanks for confirming!
I just reinstalled Quarto and Julia and redid everything and confirmed it works (not sure why it failed last time even though I updated manifest.toml).
I think this issue can be closed.

**Comment by jkrumbiegel** [comment]
Could be that an old quarto Julia server process was still running and it picked up that one after updating, so you only thought you'd tested on the correct QNB version. This is something I still want to improve, ideally the server would restart when a new QNB version is detected.

**Comment by cderv** [comment]
Closed by #10110 following update of QuartoNotebookRunner
